Fix regression with regard to quotes in attribute values

mfwitten opened the pull request:
The following commit removed the escaping of single quotes and double quotes, in order to fix issues #21 and #56:

&nbsp;&nbsp;&nbsp;&nbsp;**commit** 57f93efc894b1ba4c30e42b549b0081185029c36  
&nbsp;&nbsp;&nbsp;&nbsp;**Date:** Tue Aug 16 00:10:41 2016 +0200

However, this broke the quoting of an attribute value that has both a single quote and a double quote. This patch series re-introduces the escaping of quotes, but ultimately uses entities only when both a single quote and a double quote are present at the same time in an attribute value; whether single quotes or double quotes are escaped depends on which choice produces the shortest result. Unit tests are included.

I didn't find out about issue #21 or #56 until just now, so it might be worthwhile to include tests for those situations before proceeding.

Ede123 commented:
Thank you for this PR Michael, looks very promising overall!

First things first:<br/>
Would it be too much to ask to split this PR? I'd be happy to merge some changes right away (e.g. 0df46a6cbc84023fcc575ab7ed504cf0608a0801, 95b1df843f9013a51fa62171d695e187269486c7), but I have some minor questions and modifications for others which I'd like to resolve first.
If you prefer to keep the PR intact at this point I understand but I think it makes reviewing and adjusting the individual commits a lot harder for both of us...

Regarding the central change:<br/>
Could you clarify the issue the current behavior causes for you? When I implemented 57f93efc894b1ba4c30e42b549b0081185029c36 I thought about possibilities where entities in attribute names could be desired but I came to the conclusion that they're probably invalid in all typical cases (as stated in my commit message).

mfwitten commented:
I've split this pull request into a number of quasi-separate pull requests: #149, #150, #151, #152. For now, I've left out the changes to `README.md`, they can be handled another time.

They are "quasi-separate", because each one forms the foundation for the next, so as to make for fast-forward merges without the need to do any rebasing. Of course, if there is a need to make changes to one of them, then they'll have to be rebased, anyway; however, I presume #149, #150, and #151 to be non-controversial.

Ede123 commented:
Well, "quasi-separate" wasn't exactly what I had in mind (not much advantage in that... Ideally they would not depend on each other)

As a result I already messed it up (seems GitHub doesn't do fast-forward merging) so they all need to be rebased now...

Ede123 commented:
You can combine #150 and #151 if you want (those should be ready to merge), I'll have another look at #152 afterwards...

mfwitten commented:
Sorry for the confusion.

I see now that GitHub uses `--no-ff` by default, and that the only way to get a fast-forward merge from GitHub's interface is basically to use the "Rebase and merge", which I suspect is what you used.

I've closed #150, and incorporated its commits into #151.

I've also rebased #151 onto the results from merging pull request #149, and I've decided to go ahead and rebase #152 onto the new #151, because that makes more sense to me than pretending like it was written on top of some earlier ancestor.

**However**, I suspect that I didn't need to do all of this rebasing on my end, anyway; by default `git rebase` throws out empty commits (ones that make no changes), and thus you should have been able to run "Rebase and merge" on both #150 and #151 without trouble, unless GitHub does something silly like `--keep-empty`.

So, now, after running "Rebase and merge" on #151: If you like #152 as it stands, you should be able to run "Rebase and merge" on it without worrying about the stale ancestors.

Ede123 commented:
> I see now that GitHub uses --no-ff by default, and that the only way to get a fast-forward merge from GitHub's interface is basically to use the "Rebase and merge", which I suspect is what you used.

Yes, I used it for #149 and it seems to always rebase even if it would not be needed (i.e. fast-forward would be possible).

> However, I suspect that I didn't need to do all of this rebasing on my end, anyway; by default git rebase throws out empty commits (ones that make no changes), and thus you should have been able to run "Rebase and merge" on both #150 and #151 without trouble, unless GitHub does something silly like --keep-empty.

GitHub certainly does *something* as I wasn't able to merge the commits at all anymore after I initially used "Rebase and merge" on #149.<br/>
However one can merge on the command line with a fast-forward merge (which is what I did for #151), which then did not make #152 umergeable (even before you rebased it again).
